src/theory/sets/theory_sets_type_rules.cpp

```diff
@@ -30,6 +30,67 @@ namespace sets {
 
 using namespace cvc5::internal::theory::datatypes;
 
+bool isMaybeRelation(const TypeNode& tn)
+{
+  // if not a set type
+  if (!tn.isMaybeKind(Kind::SET_TYPE))
+  {
+    return false;
+  }
+  // if a concrete set type whose element type is not a tuple type
+  if (tn.isSet() && !tn[0].isMaybeKind(Kind::TUPLE_TYPE))
+  {
+    return false;
+  }
+  return true;
+}
+
+bool checkFunctionTypeFor(const Node& n,
+                          const TypeNode& functionType,
+                          const TypeNode& setType,
+                          std::ostream* errOut)
+{
+  // get the element type of the second argument, if it exists
+  TypeNode elementType;
+  if (setType.isSet())
+  {
+    elementType = setType.getSetElementType();
+  }
+  if (!functionType.isMaybeKind(Kind::FUNCTION_TYPE))
+  {
+    if (errOut)
+    {
+      (*errOut) << "Operator " << n.getKind()
+                << " expects a function as a first argument. "
+                << "Found a term of type '" << functionType << "'.";
+    }
+    return false;
+  }
+  // note that if functionType is abstract, we don't check whether it
+  // matches the argument.
+  if (functionType.isFunction())
+  {
+    std::vector<TypeNode> argTypes = functionType.getArgTypes();
+    if (!(argTypes.size() == 1
+          && (elementType.isNull() || argTypes[0].isComparableTo(elementType))))
+    {
+      if (errOut)
+      {
+        (*errOut) << "Operator " << n.getKind()
+                  << " expects a function whose type is comparable to the "
```

Comment:
I prefer to print `n` here because without it, it is hard to find in the source file.

Reply:
It is generally not safe to print nodes within type rules, since the printer for nodes may require computing the type of the node it is printing.

I think we actually violate this in some cases; still I think its best to avoid these issues and perhaps find a way to print the term that failed to type check in a more central place.

Reply:
I agree. But technically it should be possible to print the parsed text as it is.

Reply:
Which parsed text do you mean?  This can be called on any internal Node regardless of whether it came from parsing.

Reply:
Right. I only focused on smtlib.